### PR TITLE
fix: Rewrite the Linux module finder

### PR DESCRIPTION
**Pull request by Swatinem**
It now works in memory, without requiring to mmap the libraries again,
which should make it work correctly on android when loading libraries
directly from apk or appbundle files.

I think the problem with the old memory-unsafe module finder came from
some of these mappings:

```
7f8cd3467000-7f8cd3475000 r--p 00000000 00:1c 7597971                    /usr/lib/libcurl.so.4.7.0
7f8cd3475000-7f8cd34da000 r-xp 0000e000 00:1c 7597971                    /usr/lib/libcurl.so.4.7.0
7f8cd34da000-7f8cd34f6000 r--p 00073000 00:1c 7597971                    /usr/lib/libcurl.so.4.7.0
7f8cd34f6000-7f8cd34f7000 ---p 0008f000 00:1c 7597971                    /usr/lib/libcurl.so.4.7.0
7f8cd34f7000-7f8cd34fa000 r--p 0008f000 00:1c 7597971                    /usr/lib/libcurl.so.4.7.0
7f8cd34fa000-7f8cd34fc000 rw-p 00092000 00:1c 7597971                    /usr/lib/libcurl.so.4.7.0
```

NOTE that the 4th mapping is **not** readable, and it has the same
*offset in file* as the 5th mapping. This essentially means that we have
a gap in the address space, and trying to access it will lead to
segfaults.

The new code deals primarily with file offsets, and translates them
safely to in-memory addresses, making sure that we actually can read
whatever we are trying to read.

### CODE REVIEW

**Comment by Swatinem** [comment]
Been testing a bit with android. this is what I get when extracting native libs:
```
73786be13000-73786be17000 r-xp 00000000 fd:05 40972                      /data/app/~~WAc02Q9FOcPyBtElEJRP4A==/io.sentry.samples.android-nRHo8bDX2EhOkbXOm78y6A==/lib/x86_64/libsentry-android.so
73786be17000-73786be18000 r--p 00003000 fd:05 40972                      /data/app/~~WAc02Q9FOcPyBtElEJRP4A==/io.sentry.samples.android-nRHo8bDX2EhOkbXOm78y6A==/lib/x86_64/libsentry-android.so
73786be48000-73786bfb1000 r-xp 00000000 fd:05 40973                      /data/app/~~WAc02Q9FOcPyBtElEJRP4A==/io.sentry.samples.android-nRHo8bDX2EhOkbXOm78y6A==/lib/x86_64/libsentry.so
73786bfb1000-73786bfb2000 ---p 00000000 00:00 0 
73786bfb2000-73786bfbf000 r--p 00169000 fd:05 40973                      /data/app/~~WAc02Q9FOcPyBtElEJRP4A==/io.sentry.samples.android-nRHo8bDX2EhOkbXOm78y6A==/lib/x86_64/libsentry.so
73786bfbf000-73786bfc0000 rw-p 00176000 fd:05 40973                      /data/app/~~WAc02Q9FOcPyBtElEJRP4A==/io.sentry.samples.android-nRHo8bDX2EhOkbXOm78y6A==/lib/x86_64/libsentry.so
```

And this is with non-extracted native libs:
```
73786bf14000-73786bf18000 r-xp 00003000 fd:05 40965                      /data/app/~~qRRavra1M6p7dw5dff3zzA==/io.sentry.samples.android-k0fTZ-qmMx5lBAjqLvP6_g==/base.apk
73786bf18000-73786bf19000 r--p 00006000 fd:05 40965                      /data/app/~~qRRavra1M6p7dw5dff3zzA==/io.sentry.samples.android-k0fTZ-qmMx5lBAjqLvP6_g==/base.apk
73786bf20000-73786bf21000 ---p 00000000 00:00 0 
73786bf21000-73786bf42000 rw-p 00000000 00:00 0                          [anon:scudo:secondary]
73786bf42000-73786bf43000 ---p 00000000 00:00 0 
73786bf43000-73786c0ac000 r-xp 00432000 fd:05 40965                      /data/app/~~qRRavra1M6p7dw5dff3zzA==/io.sentry.samples.android-k0fTZ-qmMx5lBAjqLvP6_g==/base.apk
73786c0ac000-73786c0ad000 ---p 00000000 00:00 0 
73786c0ad000-73786c0ba000 r--p 0059b000 fd:05 40965                      /data/app/~~qRRavra1M6p7dw5dff3zzA==/io.sentry.samples.android-k0fTZ-qmMx5lBAjqLvP6_g==/base.apk
73786c0ba000-73786c0bb000 rw-p 005a8000 fd:05 40965                      /data/app/~~qRRavra1M6p7dw5dff3zzA==/io.sentry.samples.android-k0fTZ-qmMx5lBAjqLvP6_g==/base.apk
```

so there is just a bit of executable code loaded from the middle of the apk, as expected. But we don’t have any names to work with here either.

**Comment by Swatinem** [comment]
I updated this and added a few more cases to handle:

```
737b5570d000-737b5570e000 r--p 00000000 07:70 34                         /apex/com.android.runtime/lib64/bionic/libdl.so
737b5570e000-737b5570f000 r-xp 00000000 07:70 34                         /apex/com.android.runtime/lib64/bionic/libdl.so
737b5570f000-737b55710000 r--p 00000000 07:70 34                         /apex/com.android.runtime/lib64/bionic/libdl.so
```
Here, we have different mappings for the *same* file offset. For this case, we will prefer the `x` (executable) mapping.

For in-apk libraries, we consider the file offset as well when trying to parse the ELF and extract the code/debug-id, which seems to work.

File: src/modulefinder/sentry_modulefinder_linux.c [hunk 1]
```diff
@@ -10,25 +10,80 @@
...
+ * Checks that `start_offset` + `size` is a valid contiguous mapping in the
+ * mapped regions, and returns the translated pointer corresponding to
+ * `start_offset`.
+ */
+void *
+sentry__module_get_addr(
+    const sentry_module_t *module, uint64_t start_offset, uint64_t size)
+{
+    uint64_t addr = 0;
+    uint64_t addr_end = UINT64_MAX;
```
[the remark this answers is not in the document]
Comment (Swatinem) [reply src/modulefinder/sentry_modulefinder_linux.c:38]: We only have a max number of 5 mappings, binary search is probably overkill there and maybe even worse.

**Comment by Swatinem** [comment]
Current blocker:

Figure out which start/size we want to report for the case of duplicated mappings as shown in https://github.com/getsentry/sentry-native/pull/431#issuecomment-811843019

File: src/modulefinder/sentry_modulefinder_linux.h [hunk 2]
```diff
@@ -4,26 +4,45 @@
...
+    uint64_t size; // size of this mapping
+    uint64_t addr; // addr in memory of the mapping
+} sentry_mapped_region_t;
 
-#if SENTRY_UNITTEST
-bool sentry__mmap_file(sentry_mmap_t *mapping, const char *path);
-void sentry__mmap_close(sentry_mmap_t *mapping);
+typedef struct {
+    sentry_slice_t file;
+    sentry_mapped_region_t mappings[5];
```
Comment (flub) [inline src/modulefinder/sentry_modulefinder_linux.h:26]: `5` should be `SENTRY_MAX_MAPPINGS` instead?

**Review by flub — APPROVED** [review]

**Pull request merged**